pennylane/estimator/compact_hamiltonian.py

```diff
@@ -178,3 +180,194 @@ def __post_init__(self):
+
+        >>> num_steps, order = (10, 2)
+        >>> res = qre.estimate(qre.TrotterPauli(pauli_ham, num_steps, order))
+        >>> print(res)
+        --- Resources: ---
+         Total wires: 40
+           algorithmic wires: 40
+           allocated wires: 0
+             zero state: 0
+             any state: 0
+         Total gates : 5.266E+4
+           'T': 4.708E+4,
+           'CNOT': 1.260E+3,
+           'Z': 600,
+           'S': 1.200E+3,
+           'Hadamard': 2.520E+3
+
+    """
+
+    def __init__(
+        self,
+        num_qubits: int,
+        pauli_terms: dict | Iterable[dict],
+        one_norm: float | None = None,
+    ):
+        self._num_qubits = num_qubits
+        if one_norm is not None and not (isinstance(one_norm, (float, int)) and one_norm >= 0):
+            raise ValueError(
+                f"one_norm, if provided, must be a positive float or integer. Instead received {one_norm}"
+            )
+
+        if isinstance(pauli_terms, dict):
+            _validate_pauli_terms(pauli_terms)
+        else:
+            for group in pauli_terms:
+                _validate_pauli_terms(group)
+
+        self._one_norm = one_norm
+        self._pauli_terms = pauli_terms
+
+    def __repr__(self):
+        """The repr dunder method for the PauliHamiltonian class."""
+        return f"PauliHamiltonian(num_qubits={self.num_qubits}, one_norm={self.one_norm}, pauli_terms={self.pauli_terms})"
+
+    def __eq__(self, other: "PauliHamiltonian"):
+        """Check if two PauliHamiltonians are identical"""
+        return all(
+            (
+                self._num_qubits == other._num_qubits,
+                self._pauli_terms == other._pauli_terms,
+                self._one_norm == other._one_norm,
+            )
+        )
+
+    def __hash__(self):
+        """Hash function for the compact Hamiltonian representation"""
+        if isinstance(self._pauli_terms, dict):
+            hashable_param = _sort_and_freeze(self._pauli_terms)
+        else:
+            hashable_param = tuple(_sort_and_freeze(group) for group in self._pauli_terms)
+
+        hashable_params = (
+            self._num_qubits,
+            hashable_param,
+            self._one_norm,
+        )
+        return hash(hashable_params)
+
+    @property
+    def num_qubits(self):
+        """The number of qubits the Hamiltonian acts on"""
+        return self._num_qubits
+
+    @property
+    def one_norm(self):
+        """The one-norm of the Hamiltonian."""
+        return self._one_norm
+
+    @property
+    def pauli_terms(self):
+        """A dictionary representing the distribution of Pauli words in the Hamiltonian"""
+        return copy.deepcopy(self._pauli_terms)
+
+
+def _sort_and_freeze(pauli_terms: dict) -> tuple[tuple]:
+    """Map a dictionary into a sorted and hashable tuple"""
+    return tuple((k, pauli_terms[k]) for k in sorted(pauli_terms))
+
+
+def _validate_pauli_terms(pauli_terms: dict) -> bool:
```

Comment:
```suggestion
def _validate_pauli_terms(pauli_terms: dict | Iterable[dict]) -> bool:
```

Reply:
Actually this private method accepts just a dict.